coordinator/src/orderbook/mod.rs

```diff
@@ -30,120 +41,317 @@ pub mod websocket;
                 return a.timestamp.cmp(&b.timestamp);
             }
-            // Ascending order.
-            a.price.cmp(&b.price)
-        });
+            match direction {
+                // Descending order.
+                commons::Direction::Short => b.price.cmp(&a.price),
+                // Ascending order.
+                commons::Direction::Long => a.price.cmp(&b.price),
+            }
+        })
     }
 }
 
 impl Orderbook {
-    pub fn new(pool: Pool<ConnectionManager<PgConnection>>) -> Self {
+    pub fn new(
+        pool: Pool<ConnectionManager<PgConnection>>,
+        order_matching_fee_rate: Decimal,
+    ) -> Self {
         Self {
             pool,
-            shorts: Shorts(vec![]),
-            longs: Longs(vec![]),
+            shorts: OrderbookSide::new(),
+            longs: OrderbookSide::new(),
+            order_matching_fee_rate,
         }
     }
 
+    /// Initializes the orderbook with non expired open limit orders.
     async fn initialize(&mut self) -> Result<()> {
         let all_orders = spawn_blocking({
             let mut conn = self.pool.clone().get()?;
             move || {
+                // TODO(holzeis): join with trades to get partially matched orders.
                 let all_orders =
                     orders::get_all_orders(&mut conn, OrderType::Limit, OrderState::Open, true)?;
                 anyhow::Ok(all_orders)
             }
         })
         .await??;
 
+        // TODO(holzeis): maybe consider batch adding all orders and sort afterwards instead of
+        // adding every order individually.
         for order in all_orders {
             match order.direction {
-                commons::Direction::Short => self.shorts.add(order.clone()),
-                commons::Direction::Long => self.longs.add(order.clone()),
+                commons::Direction::Short => self.shorts.add_order(order),
+                commons::Direction::Long => self.longs.add_order(order),
             }
         }
 
         Ok(())
     }
 
-    async fn add_order(
+    /// Adds a limit order to the orderbook.
+    async fn add_limit_order(&mut self, new_order: NewLimitOrder) -> Result<Message> {
+        let order = spawn_blocking({
+            let mut conn = self.pool.clone().get()?;
+            move || {
+                let order = orders::insert_limit_order(&mut conn, new_order, OrderReason::Manual)
+                    .map_err(|e| anyhow!(e))
+                    .context("Failed to insert new order into DB")?;
+
+                anyhow::Ok(order)
+            }
+        })
+        .await??;
+
+        match order.direction {
+            commons::Direction::Short => self.shorts.add_order(order),
+            commons::Direction::Long => self.longs.add_order(order),
+        }
+        Ok(Message::NewOrder(order))
+    }
+
+    /// Matches a market order against the orderbook. Will fail if the market order can't be fully
+    /// matched.
+    async fn match_market_order(
```

Comment:
🤔 I'm not sure any of these methods need to be `async`.

Reply:
you mean if we aren't waiting for the blocking database transactions? if you you might be right.